### ISSUE TITLE
Implement a way to define lists of components

### ISSUE THREAD

**Issue by tigris** [mdub/config_mapper#6]
We would like a way to define lists. In my head, the config would look like this:

```yaml
people:
-
  name: danial pearce
  awesomeness: 10
  preferences:
    drink: beer
    food: burgers
-
  name: fred
  awesomeness: 2
  preferences:
    drink: water
```

And the model might look something like this:

```ruby
class Config < ConfigMapper::ConfigStruct
  component_list :people do
    attribute :name, String
    attribute :awesomeness, Integer, default: 0
    component_dict :preferences do
      attribute :drink, String
      attribute :food, String, default: "parmas"
    end
  end
end
```

Which would map to PORs like so:

```json
{
  "people": [
    { "name": "danial pearce", "awesomeness": 10, "preferences": { "drink": "beer", "food": "burgers" } },
    { "name": "fred", "awesomeness": 2, "preferences": { "drink": "wine", "food": "parmas" } },
  ]
}
```

I think doing the above is currently possible by creating a `Types::People`, and using `component :people, type: Types::People` although I've not yet tried. I think the above would be a nicer solution in the long run, as the entire config definition would be in 1 place instead of spread out across multiple types, where the "type" also has other stuff mingled in with it and doesn't read quite as nice as a simple config definition.

**Comment by tigris** [comment]
FWIW, i'm investigating the code required for this and open to suggestions for the PR.

**Comment by mdub** [comment]
Hey @tigris - I'm just back from vacation.

I'm pretty sure this is already possible, using `component_dict`, e.g.

```ruby
class Config < ConfigMapper::ConfigStruct
  component_dict :people do
    attribute :name, String
    attribute :awesomeness, Integer, default: 0
    component_dict :preferences do
      attribute :drink, String
      attribute :food, String, default: "parmas"
    end
  end
end
```

The `people` attribute would end up looking like a Map (keyed by Integer array indices), so you'd need to call `people.values` to get the details. Would that work?

**Comment by tigris** [comment]
G'day @mdub - I'm just leaving for vacation! 😄 

That does indeed work. Although I had to monkey patch in the `:values` delegator to `ConfigDict` as it's not currently there.

However, it doesn't provide a way to force that. I.e. you can still _name_ each entry of the dict in the source data whatever you want. I guess that's not the worst. I just felt something like `component_list` felt more natural for when you want to force a list. I guess you could alias them. 

The other upside of the `component_list` is not require a `.values` to get them, calling the `.people` method returns the array as is.

Lastly, the internals of the `component_dict` have to be a complex object (by default, a `ConfigDict`). Can you think of an easy way to force/validate a list of basic objects, such as Integer etc, the same way you can for an `attribute`?

At the moment you can do `attribute :things, Array` and get at `.things` pretty easily, but whatever source data is shoved into that array is simply returned as is, unvalidated. If `component_list` existed, in block form it could do the `.people` example above, where the block defines a complex dictionary. But using an inline syntax like `component_list :things, type: Integer` could be nice syntactic sugar for validating an array of integers.

**Comment by tigris** [comment]
FWIW, using `dry-rb` to try and use `component :things, type: Type::Array.of(String)` didn't work.

I just tried to implement my own `Type::Array.of(klass)` to use as a `component`, but I failed. I'll keep the code handy for now, but I fly out in the morning, so it's nap time.

**Comment by mdub** [comment]
Happy napping and holidaying, @tigris . 

You've convinced me that it's worth having a specific `component_list` method in the DSL, with a `ConfigList` container behind it.

It's true that values in a `component_dict` must be components (complex objects) ... and the same would be true for `component_list`. There's probably value in supporting lists/maps of "basic objects", but let's tackle that separately (i.e. new issue, different PR). I imagine adding `list` and `dict` methods to the DSL to support that use-case.

**Comment by mdub** [comment]
Done in #7 (thanks @tigris). Released as 1.7.0.
